corehq/apps/users/validation.py

```diff
@@ -0,0 +1,47 @@
+from django.core.exceptions import ValidationError
+from django.core.validators import validate_email
+
+from corehq.apps.users.dbaccessors import user_exists
+from corehq.apps.users.exceptions import (
+    InvalidDomainException,
+    InvalidUsernameException,
+    ReservedUsernameException,
+    UsernameAlreadyExists,
+)
+from corehq.apps.users.util import format_username
+from corehq.apps.users.views.mobile import BAD_MOBILE_USERNAME_REGEX
+
+
+def validate_mobile_username(username, domain):
+    if username is None:
+        raise InvalidUsernameException
+    if domain is None:
+        raise InvalidDomainException
+
+    _check_for_reserved_usernames(username)
+    username_as_email = format_username(username, domain)
+    _ensure_valid_username(username_as_email)
+    _ensure_username_is_available(username_as_email)
+    return username_as_email
+
+
+def _check_for_reserved_usernames(username):
+    reserved_usernames = ['admin', 'demo_user']
+    if username in reserved_usernames:
+        raise ReservedUsernameException
```

Comment:
It's good practice to pass the bad value into the exception in case it is needed later. For example if there was code that was validating a list of usernames.
```suggestion
        raise ReservedUsernameException(username)
```

Reply:
See See e9bb80ded2dd47c3a2f1c608fe2545524ed91b59